[Bug]: Container logging exits prematurely with error

### Testcontainers version

0.29.1

### Using the latest Testcontainers version?

Yes

### Host OS

Linux

### Host arch

x86

### Go version

1.22

### Docker version

```shell
(NOTE: using podman, not docker)

Version:      3.4.4
API Version:  3.4.4
Go Version:   go1.18.1
Built:        Wed Dec 31 19:00:00 1969
OS/Arch:      linux/amd64
```


### Docker info

```shell
host:
  arch: amd64
  buildahVersion: 1.23.1
  cgroupControllers:
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: 'conmon: /usr/bin/conmon'
    path: /usr/bin/conmon
    version: 'conmon version 2.0.25, commit: unknown'
  cpus: 32
  distribution:
    codename: jammy
    distribution: ubuntu
    version: "22.04"
  eventLogger: journald
  hostname: noah-workstation
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 2109978069
      size: 1
    - container_id: 1
      host_id: 10000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 2109978069
      size: 1
    - container_id: 1
      host_id: 10000
      size: 65536
  kernel: 6.1.0-1033-oem
  linkmode: dynamic
  logDriver: journald
  memFree: 101372432384
  memTotal: 134160068608
  ociRuntime:
    name: crun
    package: 'crun: /usr/bin/crun'
    path: /usr/bin/crun
    version: |-
      crun version 0.17
      commit: 0e9229ae34caaebcb86f1fde18de3acaf18c6d9a
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/2109978069/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: 'slirp4netns: /usr/bin/slirp4netns'
    version: |-
      slirp4netns version 1.0.1
      commit: 6a7b16babc95b6a3056b33fb45b74a6f62262dd4
      libslirp: 4.6.1
  swapFree: 8589930496
  swapTotal: 8589930496
  uptime: 363h 0m 53.36s (Approximately 15.12 days)
plugins:
  log:
  - k8s-file
  - none
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries: {}
store:
  configFile: /home/noah/.config/containers/storage.conf
  containerStore:
    number: 4
    paused: 0
    running: 0
    stopped: 4
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/noah/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 25
  runRoot: /run/user/2109978069/containers
  volumePath: /home/noah/.local/share/containers/storage/volumes
version:
  APIVersion: 3.4.4
  Built: 0
  BuiltTime: Wed Dec 31 19:00:00 1969
  GitCommit: ""
  GoVersion: go1.18.1
  OsArch: linux/amd64
  Version: 3.4.4
```


### What happened?

I've found that `LogConsumer`s sometimes stop receiving logs prematurely and dump an error. 

I've built a minimal reproducible example of the specific issue. It consists of a very small container image based on [`distroless/base`](https://github.com/GoogleContainerTools/distroless). The container executes the following program:

```go
package main

import (
	"log"
	"time"
)

func main() {
	i := 0
	for {
		log.Printf("tick %d", i)
		i += 1
		time.Sleep(1 * time.Second)
	}
}
```

I've attached a prebuilt container to the issue as a tarball.

In order to reproduce the issue, first load the container tarball like so:

```
gunzip container.tar.gz
podman load -i container.tar
```

And then compile and run the following test:

```go
package container_test

import (
	"context"
	"fmt"
	"testing"
	"time"

	tc "github.com/testcontainers/testcontainers-go"
)

type StdoutLogConsumer struct{}

func (lc *StdoutLogConsumer) Accept(l tc.Log) {
	fmt.Printf("%s", l.Content)
}

func TestLogIssue(t *testing.T) {
	req := tc.ContainerRequest{
		Image: "localhost/log_repro:latest",
		LogConsumerCfg: &tc.LogConsumerConfig{
			Consumers: []tc.LogConsumer{&StdoutLogConsumer{}},
		},
	}

	ctx := context.Background()
	container, err := tc.GenericContainer(ctx, tc.GenericContainerRequest{
		ContainerRequest: req,
		Started:          true,
	})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		if err := container.Terminate(ctx); err != nil {
			t.Fatal(err)
		}
	})

	time.Sleep(10 * time.Second)

	fmt.Println("test over")
}
```

I'd expect this test to print container logs showing 9-10 ticks and exit cleanly. Instead, it consistently stops displaying output after 6 ticks, and always outputs an error message:

```
2024/03/06 19:21:03 github.com/testcontainers/testcontainers-go - Connected to docker:
  Server Version: 3.4.4
  API Version: 1.40
  Operating System: ubuntu
  Total Memory: 127945 MB
  Resolved Docker Host: unix:///run/user/2109978069/podman/podman.sock
  Resolved Docker Socket Path: /run/user/2109978069/podman/podman.sock
  Test SessionID: c7ce1db5c519f0188cc6943c86f8faccf62034e6b4d4c1ab5bf6f8ff896ecf3b
  Test ProcessID: 36d03253-4e13-441e-8149-88e46fb4fa15
2024/03/06 19:21:03 🐳 Creating container for image testcontainers/ryuk:0.6.0
2024/03/06 19:21:03 ✅ Container created: 15f55739196d
2024/03/06 19:21:03 🐳 Starting container: 15f55739196d
2024/03/06 19:21:03 ✅ Container started: 15f55739196d
2024/03/06 19:21:03 🚧 Waiting for container id 15f55739196d image: testcontainers/ryuk:0.6.0. Waiting for: &{Port:8080/tcp timeout:<nil> PollInterval:100ms}
2024/03/06 19:21:03 🔔 Container is ready: 15f55739196d
2024/03/06 19:21:03 🐳 Creating container for image localhost/log_repro:latest
2024/03/06 19:21:03 ✅ Container created: f8583fd38a61
2024/03/06 19:21:03 🐳 Starting container: f8583fd38a61
2024/03/06 19:21:03 ✅ Container started: f8583fd38a61
2024/03/06 19:21:03 🔔 Container is ready: f8583fd38a61
2024/03/06 19:21:03 tick 0
2024/03/06 19:21:04 tick 1
2024/03/06 19:21:05 tick 2
2024/03/06 19:21:06 tick 3
2024/03/06 19:21:07 tick 4
2024/03/06 19:21:08 tick 5
container log error: EOF. Stopping log consumer: Headers out of sync2024/03/06 19:21:13 🐳 Terminating container: f8583fd38a61
test over
2024/03/06 19:21:14 🚫 Container terminated: f8583fd38a61
PASS
```


I did notice some interesting things about this while trying to debug, but I'm not sure what to make of them. In the main loop of `docker.startLogProduction()`, it seems like the `"use of closed network connection"` error block consistently triggers before the EOF error is triggered. In addition, it seems like attempting to restart the connection when EOF is received allows the log consumer to get a few more logs, e.g.:

```go
//_, _ = fmt.Fprintf(os.Stderr, "container log error: %+v. %s", err, logStoppedForOutOfSyncMessage)
//// if we would continue here, the next header-read will result into random data...
//return

now := time.Now()
since = fmt.Sprintf("%d.%09d", now.Unix(), int64(now.Nanosecond()))
goto BEGIN
```


### Relevant log output

_No response_

### Additional information

[container.tar.gz](https://github.com/testcontainers/testcontainers-go/files/14514644/container.tar.gz)

Alright, I was able to reproduce this with a pretty simple example. I've attached a new image tar, this container runs a binary that does nothing and simply busy-loops forever. 

The test code itself is also fairly minimal:

```go
package container_test

import (
	"context"
	"fmt"
	"testing"

	tc "github.com/testcontainers/testcontainers-go"
)

type StdoutLogConsumer struct{}

func (lc *StdoutLogConsumer) Accept(l tc.Log) {
	fmt.Printf("%s", l.Content)
}

func TestLogIssue(t *testing.T) {
	req := tc.ContainerRequest{
		Image: "localhost/log_repro:latest",
		LogConsumerCfg: &tc.LogConsumerConfig{
			Consumers: []tc.LogConsumer{&StdoutLogConsumer{}},
		},
	}

	ctx := context.Background()

	container, err := tc.GenericContainer(ctx, tc.GenericContainerRequest{
		ContainerRequest: req,
		Started:          true,
	})
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		if err := container.Terminate(ctx); err != nil {
			t.Fatal(err)
		}
	})

	fmt.Println("test over")
}
```

The output I get:

```
2024/03/15 20:47:46 github.com/testcontainers/testcontainers-go - Connected to docker:
  Server Version: 4.9.3
  API Version: 1.41
  Operating System: ubuntu
  Total Memory: 127946 MB
  Resolved Docker Host: unix:///run/user/2109978069/podman/podman.sock
  Resolved Docker Socket Path: /run/user/2109978069/podman/podman.sock
  Test SessionID: 61f8be8f5fb4a6f37c09cd82a69ea05022e3a674641b2add09d5c5593b17f0c2
  Test ProcessID: 9ffc22d9-7015-4555-be43-c7ebd115a846
2024/03/15 20:47:46 🐳 Creating container for image testcontainers/ryuk:0.6.0
2024/03/15 20:47:46 ✅ Container created: 1c9f5ce3ad48
2024/03/15 20:47:46 🐳 Starting container: 1c9f5ce3ad48
2024/03/15 20:47:46 ✅ Container started: 1c9f5ce3ad48
2024/03/15 20:47:46 🚧 Waiting for container id 1c9f5ce3ad48 image: testcontainers/ryuk:0.6.0. Waiting for: &{Port:8080/tcp timeout:<nil> PollInterval:100ms}
2024/03/15 20:47:46 🔔 Container is ready: 1c9f5ce3ad48
2024/03/15 20:47:46 🐳 Creating container for image localhost/log_repro:latest
2024/03/15 20:47:46 ✅ Container created: b3b72d3709fe
2024/03/15 20:47:46 🐳 Starting container: b3b72d3709fe
2024/03/15 20:47:46 ✅ Container started: b3b72d3709fe
2024/03/15 20:47:46 🔔 Container is ready: b3b72d3709fe
test over
2024/03/15 20:47:46 🐳 Terminating container: b3b72d3709fe
--- FAIL: TestLogIssue (6.78s)
    container_test.go:36: Get "http://%2Frun%2Fuser%2F2109978069%2Fpodman%2Fpodman.sock/v1.41/containers/b3b72d3709fed603ed1ddc15159307171130ac77a9981ee86610cbde2e05970f/logs?follow=1&stderr=1&stdout=1&tail=": context deadline exceeded
FAIL
```

--- a/docker.go
+++ b/docker.go
@@ -63,16 +63,26 @@ type DockerContainer struct {
 	isRunning     bool
 	imageWasBuilt bool
 	// keepBuiltImage makes Terminate not remove the image if imageWasBuilt.
-	keepBuiltImage       bool
-	provider             *DockerProvider
-	sessionID            string
-	terminationSignal    chan bool
-	consumers            []LogConsumer
-	raw                  *types.ContainerJSON
-	stopLogProductionCh  chan bool
-	logProductionDone    chan bool
-	logProductionError   chan error
-	logProductionMutex   sync.Mutex
+	keepBuiltImage     bool
+	provider           *DockerProvider
+	sessionID          string
+	terminationSignal  chan bool
+	consumers          []LogConsumer
+	raw                *types.ContainerJSON
+	logProductionError chan error
+
+	// TODO: Remove locking and wait group once the deprecated StartLogProducer and
+	// StopLogProducer have been removed and hence logging can only be started and
+	// stopped once.
+
+	// logProductionWaitGroup is used to signal when the log production has stopped.
+	// This allows stopLogProduction to safely set logProductionStop to nil.
+	logProductionWaitGroup sync.WaitGroup
+
+	// logProductionMutex protects logProductionStop channel so it can be started again.
+	logProductionMutex sync.Mutex
+	logProductionStop  chan struct{}
+
 	logProductionTimeout *time.Duration
 	logger               Logging
 	lifecycleHooks       []ContainerLifecycleHooks
@@ -675,9 +685,12 @@ func (c *DockerContainer) startLogProduction(ctx context.Context, opts ...LogPro
 		c.logProductionMutex.Lock()
 		defer c.logProductionMutex.Unlock()
 
-		if c.stopLogProductionCh != nil {
+		if c.logProductionStop != nil {
 			return errors.New("log production already started")
 		}
+
+		c.logProductionStop = make(chan struct{})
+		c.logProductionWaitGroup.Add(1)
 	}
 
 	for _, opt := range opts {
@@ -699,21 +712,12 @@ func (c *DockerContainer) startLogProduction(ctx context.Context, opts ...LogPro
 		c.logProductionTimeout = &maxLogProductionTimeout
 	}
 
-	c.stopLogProductionCh = make(chan bool)
-	c.logProductionDone = make(chan bool)
 	c.logProductionError = make(chan error, 1)
 
-	go func(stop <-chan bool, done chan<- bool, errorCh chan error) {
-		// signal the log production is done once go routine exits, this prevents race conditions around start/stop
-		// set c.stopLogProductionCh to nil so that it can be started again
+	go func() {
 		defer func() {
-			defer c.logProductionMutex.Unlock()
-			close(done)
-			close(errorCh)
-			{
-				c.logProductionMutex.Lock()
-				c.stopLogProductionCh = nil
-			}
+			close(c.logProductionError)
+			c.logProductionWaitGroup.Done()
 		}()
 
 		since := ""
@@ -731,15 +735,15 @@ func (c *DockerContainer) startLogProduction(ctx context.Context, opts ...LogPro
 
 		r, err := c.provider.client.ContainerLogs(ctx, c.GetContainerID(), options)
 		if err != nil {
-			errorCh <- err
+			c.logProductionError <- err
 			return
 		}
 		defer c.provider.Close()
 
 		for {
 			select {
-			case <-stop:
-				errorCh <- r.Close()
+			case <-c.logProductionStop:
+				c.logProductionError <- r.Close()
 				return
 			default:
 				h := make([]byte, 8)
@@ -795,7 +799,7 @@ func (c *DockerContainer) startLogProduction(ctx context.Context, opts ...LogPro
 				}
 			}
 		}
-	}(c.stopLogProductionCh, c.logProductionDone, c.logProductionError)
+	}()
 
 	return nil
 }
@@ -805,17 +809,18 @@ func (c *DockerContainer) StopLogProducer() error {
 	return c.stopLogProduction()
 }
 
-// StopLogProducer will stop the concurrent process that is reading logs
+// stopLogProduction will stop the concurrent process that is reading logs
 // and sending them to each added LogConsumer
 func (c *DockerContainer) stopLogProduction() error {
+	// TODO: Remove locking and wait group once StartLogProducer and StopLogProducer
+	// have been removed and hence logging can only be started / stopped once.
 	c.logProductionMutex.Lock()
 	defer c.logProductionMutex.Unlock()
-	if c.stopLogProductionCh != nil {
-		c.stopLogProductionCh <- true
-		// block until the log production is actually done in order to avoid strange races
-		<-c.logProductionDone
-		c.stopLogProductionCh = nil
-		c.logProductionDone = nil
+	if c.logProductionStop != nil {
+		close(c.logProductionStop)
+		c.logProductionWaitGroup.Wait()
+		// Set c.logProductionStop to nil so that it can be started again.
+		c.logProductionStop = nil
 		return <-c.logProductionError
 	}
 	return nil
@@ -1122,17 +1127,16 @@ func (p *DockerProvider) CreateContainer(ctx context.Context, req ContainerReque
 	}
 
 	c := &DockerContainer{
-		ID:                  resp.ID,
-		WaitingFor:          req.WaitingFor,
-		Image:               imageName,
-		imageWasBuilt:       req.ShouldBuildImage(),
-		keepBuiltImage:      req.ShouldKeepBuiltImage(),
-		sessionID:           core.SessionID(),
-		provider:            p,
-		terminationSignal:   termSignal,
-		stopLogProductionCh: nil,
-		logger:              p.Logger,
-		lifecycleHooks:      req.LifecycleHooks,
+		ID:                resp.ID,
+		WaitingFor:        req.WaitingFor,
+		Image:             imageName,
+		imageWasBuilt:     req.ShouldBuildImage(),
+		keepBuiltImage:    req.ShouldKeepBuiltImage(),
+		sessionID:         core.SessionID(),
+		provider:          p,
+		terminationSignal: termSignal,
+		logger:            p.Logger,
+		lifecycleHooks:    req.LifecycleHooks,
 	}
 
 	err = c.createdHook(ctx)
@@ -1225,15 +1229,14 @@ func (p *DockerProvider) ReuseOrCreateContainer(ctx context.Context, req Contain
 	}
 
 	dc := &DockerContainer{
-		ID:                  c.ID,
-		WaitingFor:          req.WaitingFor,
-		Image:               c.Image,
-		sessionID:           sessionID,
-		provider:            p,
-		terminationSignal:   termSignal,
-		stopLogProductionCh: nil,
-		logger:              p.Logger,
-		lifecycleHooks:      []ContainerLifecycleHooks{combineContainerHooks(defaultHooks, req.LifecycleHooks)},
+		ID:                c.ID,
+		WaitingFor:        req.WaitingFor,
+		Image:             c.Image,
+		sessionID:         sessionID,
+		provider:          p,
+		terminationSignal: termSignal,
+		logger:            p.Logger,
+		lifecycleHooks:    []ContainerLifecycleHooks{combineContainerHooks(defaultHooks, req.LifecycleHooks)},
 	}
 
 	err = dc.startedHook(ctx)
@@ -1545,7 +1548,6 @@ func containerFromDockerResponse(ctx context.Context, response types.Container) 
 
 	container.sessionID = core.SessionID()
 	container.consumers = []LogConsumer{}
-	container.stopLogProductionCh = nil
 	container.isRunning = response.State == "running"
 
 	// the termination signal should be obtained from the reaper

--- a/docker.go
+++ b/docker.go
@@ -274,22 +274,13 @@ func (c *DockerContainer) Terminate(ctx context.Context) error {
 
 	defer c.provider.client.Close()
 
-	err := c.terminatingHook(ctx)
-	if err != nil {
-		return err
-	}
-
-	err = c.provider.client.ContainerRemove(ctx, c.GetContainerID(), container.RemoveOptions{
-		RemoveVolumes: true,
-		Force:         true,
-	})
-	if err != nil {
-		return err
-	}
-
-	err = c.terminatedHook(ctx)
-	if err != nil {
-		return err
+	errs := []error{
+		c.terminatingHook(ctx),
+		c.provider.client.ContainerRemove(ctx, c.GetContainerID(), container.RemoveOptions{
+			RemoveVolumes: true,
+			Force:         true,
+		}),
+		c.terminatedHook(ctx),
 	}
 
 	if c.imageWasBuilt && !c.keepBuiltImage {
@@ -297,14 +288,12 @@ func (c *DockerContainer) Terminate(ctx context.Context) error {
 			Force:         true,
 			PruneChildren: true,
 		})
-		if err != nil {
-			return err
-		}
+		errs = append(errs, err)
 	}
 
 	c.sessionID = ""
 	c.isRunning = false
-	return nil
+	return errors.Join(errs...)
 }
 
 // update container raw info

--- a/lifecycle.go
+++ b/lifecycle.go
@@ -2,6 +2,7 @@ package testcontainers
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"io"
 	"strings"
@@ -225,65 +226,40 @@ var defaultReadinessHook = func() ContainerLifecycleHooks {
 
 // creatingHook is a hook that will be called before a container is created.
 func (req ContainerRequest) creatingHook(ctx context.Context) error {
-	for _, lifecycleHooks := range req.LifecycleHooks {
-		err := lifecycleHooks.Creating(ctx)(req)
-		if err != nil {
-			return err
-		}
+	errs := make([]error, len(req.LifecycleHooks))
+	for i, lifecycleHooks := range req.LifecycleHooks {
+		errs[i] = lifecycleHooks.Creating(ctx)(req)
 	}
 
-	return nil
+	return errors.Join(errs...)
 }
 
-// createdHook is a hook that will be called after a container is created
+// createdHook is a hook that will be called after a container is created.
 func (c *DockerContainer) createdHook(ctx context.Context) error {
-	for _, lifecycleHooks := range c.lifecycleHooks {
-		err := containerHookFn(ctx, lifecycleHooks.PostCreates)(c)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
+	return c.applyLifecycleHooks(ctx, false, func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook {
+		return lifecycleHooks.PostCreates
+	})
 }
 
-// startingHook is a hook that will be called before a container is started
+// startingHook is a hook that will be called before a container is started.
 func (c *DockerContainer) startingHook(ctx context.Context) error {
-	for _, lifecycleHooks := range c.lifecycleHooks {
-		err := containerHookFn(ctx, lifecycleHooks.PreStarts)(c)
-		if err != nil {
-			c.printLogs(ctx, err)
-			return err
-		}
-	}
-
-	return nil
+	return c.applyLifecycleHooks(ctx, true, func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook {
+		return lifecycleHooks.PreStarts
+	})
 }
 
-// startedHook is a hook that will be called after a container is started
+// startedHook is a hook that will be called after a container is started.
 func (c *DockerContainer) startedHook(ctx context.Context) error {
-	for _, lifecycleHooks := range c.lifecycleHooks {
-		err := containerHookFn(ctx, lifecycleHooks.PostStarts)(c)
-		if err != nil {
-			c.printLogs(ctx, err)
-			return err
-		}
-	}
-
-	return nil
+	return c.applyLifecycleHooks(ctx, true, func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook {
+		return lifecycleHooks.PostStarts
+	})
 }
 
-// readiedHook is a hook that will be called after a container is ready
+// readiedHook is a hook that will be called after a container is ready.
 func (c *DockerContainer) readiedHook(ctx context.Context) error {
-	for _, lifecycleHooks := range c.lifecycleHooks {
-		err := containerHookFn(ctx, lifecycleHooks.PostReadies)(c)
-		if err != nil {
-			c.printLogs(ctx, err)
-			return err
-		}
-	}
-
-	return nil
+	return c.applyLifecycleHooks(ctx, true, func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook {
+		return lifecycleHooks.PostReadies
+	})
 }
 
 // printLogs is a helper function that will print the logs of a Docker container
@@ -304,49 +280,47 @@ func (c *DockerContainer) printLogs(ctx context.Context, cause error) {
 	c.logger.Printf("container logs (%s):\n%s", cause, b)
 }
 
-// stoppingHook is a hook that will be called before a container is stopped
+// stoppingHook is a hook that will be called before a container is stopped.
 func (c *DockerContainer) stoppingHook(ctx context.Context) error {
-	for _, lifecycleHooks := range c.lifecycleHooks {
-		err := containerHookFn(ctx, lifecycleHooks.PreStops)(c)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
+	return c.applyLifecycleHooks(ctx, false, func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook {
+		return lifecycleHooks.PreStops
+	})
 }
 
-// stoppedHook is a hook that will be called after a container is stopped
+// stoppedHook is a hook that will be called after a container is stopped.
 func (c *DockerContainer) stoppedHook(ctx context.Context) error {
-	for _, lifecycleHooks := range c.lifecycleHooks {
-		err := containerHookFn(ctx, lifecycleHooks.PostStops)(c)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
+	return c.applyLifecycleHooks(ctx, false, func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook {
+		return lifecycleHooks.PostStops
+	})
 }
 
-// terminatingHook is a hook that will be called before a container is terminated
+// terminatingHook is a hook that will be called before a container is terminated.
 func (c *DockerContainer) terminatingHook(ctx context.Context) error {
-	for _, lifecycleHooks := range c.lifecycleHooks {
-		err := containerHookFn(ctx, lifecycleHooks.PreTerminates)(c)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
+	return c.applyLifecycleHooks(ctx, false, func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook {
+		return lifecycleHooks.PreTerminates
+	})
 }
 
-// terminatedHook is a hook that will be called after a container is terminated
+// terminatedHook is a hook that will be called after a container is terminated.
 func (c *DockerContainer) terminatedHook(ctx context.Context) error {
-	for _, lifecycleHooks := range c.lifecycleHooks {
-		err := containerHookFn(ctx, lifecycleHooks.PostTerminates)(c)
-		if err != nil {
-			return err
+	return c.applyLifecycleHooks(ctx, false, func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook {
+		return lifecycleHooks.PostTerminates
+	})
+}
+
+// applyLifecycleHooks applies all lifecycle hooks reporting the container logs on error if logError is true.
+func (c *DockerContainer) applyLifecycleHooks(ctx context.Context, logError bool, hooks func(lifecycleHooks ContainerLifecycleHooks) []ContainerHook) error {
+	errs := make([]error, len(c.lifecycleHooks))
+	for i, lifecycleHooks := range c.lifecycleHooks {
+		errs[i] = containerHookFn(ctx, hooks(lifecycleHooks))(c)
+	}
+
+	if err := errors.Join(errs...); err != nil {
+		if logError {
+			c.printLogs(ctx, err)
 		}
+
+		return err
 	}
 
 	return nil
@@ -369,13 +343,12 @@ func (c ContainerLifecycleHooks) Creating(ctx context.Context) func(req Containe
 // container lifecycle hooks. The created function will iterate over all the hooks and call them one by one.
 func containerHookFn(ctx context.Context, containerHook []ContainerHook) func(container Container) error {
 	return func(container Container) error {
-		for _, hook := range containerHook {
-			if err := hook(ctx, container); err != nil {
-				return err
-			}
+		errs := make([]error, len(containerHook))
+		for i, hook := range containerHook {
+			errs[i] = hook(ctx, container)
 		}
 
-		return nil
+		return errors.Join(errs...)
 	}
 }
 

--- a/logconsumer_test.go
+++ b/logconsumer_test.go
@@ -9,6 +9,7 @@ import (
 	"net/http"
 	"os"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
@@ -23,8 +24,9 @@ import (
 const lastMessage = "DONE"
 
 type TestLogConsumer struct {
-	Msgs []string
-	Done chan bool
+	mtx  sync.Mutex
+	msgs []string
+	Done chan struct{}
 
 	// Accepted provides a blocking way of ensuring the logs messages have been consumed.
 	// This allows for proper synchronization during Test_StartStop in particular.
@@ -35,11 +37,21 @@ type TestLogConsumer struct {
 func (g *TestLogConsumer) Accept(l Log) {
 	s := string(l.Content)
 	if s == fmt.Sprintf("echo %s\n", lastMessage) {
-		g.Done <- true
+		close(g.Done)
 		return
 	}
 	g.Accepted <- s
-	g.Msgs = append(g.Msgs, s)
+
+	g.mtx.Lock()
+	defer g.mtx.Unlock()
+	g.msgs = append(g.msgs, s)
+}
+
+func (g *TestLogConsumer) Msgs() []string {
+	g.mtx.Lock()
+	defer g.mtx.Unlock()
+
+	return g.msgs
 }
 
 // devNullAcceptorChan returns string channel that essentially sends all strings to dev null
@@ -57,8 +69,8 @@ func Test_LogConsumerGetsCalled(t *testing.T) {
 	ctx := context.Background()
 
 	g := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 
@@ -100,7 +112,7 @@ func Test_LogConsumerGetsCalled(t *testing.T) {
 		t.Fatal("never received final log message")
 	}
 
-	assert.Equal(t, []string{"ready\n", "echo hello\n", "echo there\n"}, g.Msgs)
+	assert.Equal(t, []string{"ready\n", "echo hello\n", "echo there\n"}, g.Msgs())
 
 	terminateContainerOnEnd(t, ctx, c)
 }
@@ -172,13 +184,13 @@ func Test_MultipleLogConsumers(t *testing.T) {
 	ctx := context.Background()
 
 	first := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 	second := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 
@@ -214,13 +226,13 @@ func Test_MultipleLogConsumers(t *testing.T) {
 	<-first.Done
 	<-second.Done
 
-	assert.Equal(t, []string{"ready\n", "echo mlem\n"}, first.Msgs)
-	assert.Equal(t, []string{"ready\n", "echo mlem\n"}, second.Msgs)
+	assert.Equal(t, []string{"ready\n", "echo mlem\n"}, first.Msgs())
+	assert.Equal(t, []string{"ready\n", "echo mlem\n"}, second.Msgs())
 	require.NoError(t, c.Terminate(ctx))
 }
 
 func TestContainerLogWithErrClosed(t *testing.T) {
-	if os.Getenv("XDG_RUNTIME_DIR") != "" {
+	if os.Getenv("GITHUB_RUN_ID") != "" {
 		t.Skip("Skipping as flaky on GitHub Actions, Please see https://github.com/testcontainers/testcontainers-go/issues/1924")
 	}
 
@@ -290,8 +302,8 @@ func TestContainerLogWithErrClosed(t *testing.T) {
 	}
 
 	consumer := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 
@@ -317,7 +329,7 @@ func TestContainerLogWithErrClosed(t *testing.T) {
 
 	// Gather the initial container logs
 	time.Sleep(time.Second * 1)
-	existingLogs := len(consumer.Msgs)
+	existingLogs := len(consumer.Msgs())
 
 	hitNginx := func() {
 		i, _, err := dind.Exec(ctx, []string{"wget", "--spider", "localhost:" + port.Port()})
@@ -328,10 +340,11 @@ func TestContainerLogWithErrClosed(t *testing.T) {
 
 	hitNginx()
 	time.Sleep(time.Second * 1)
-	if len(consumer.Msgs)-existingLogs != 1 {
-		t.Fatalf("logConsumer should have 1 new log message, instead has: %v", consumer.Msgs[existingLogs:])
+	msgs := consumer.Msgs()
+	if len(msgs)-existingLogs != 1 {
+		t.Fatalf("logConsumer should have 1 new log message, instead has: %v", msgs[existingLogs:])
 	}
-	existingLogs = len(consumer.Msgs)
+	existingLogs = len(consumer.Msgs())
 
 	iptableArgs := []string{
 		"INPUT", "-p", "tcp", "--dport", "2375",
@@ -351,10 +364,11 @@ func TestContainerLogWithErrClosed(t *testing.T) {
 	hitNginx()
 	hitNginx()
 	time.Sleep(time.Second * 1)
-	if len(consumer.Msgs)-existingLogs != 2 {
+	msgs = consumer.Msgs()
+	if len(msgs)-existingLogs != 2 {
 		t.Fatalf(
 			"LogConsumer should have 2 new log messages after detecting closed connection and"+
-				" re-requesting logs. Instead has:\n%s", consumer.Msgs[existingLogs:],
+				" re-requesting logs. Instead has:\n%s", msgs[existingLogs:],
 		)
 	}
 }
@@ -389,8 +403,8 @@ func TestContainerLogsShouldBeWithoutStreamHeader(t *testing.T) {
 func TestContainerLogsEnableAtStart(t *testing.T) {
 	ctx := context.Background()
 	g := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 
@@ -434,7 +448,7 @@ func TestContainerLogsEnableAtStart(t *testing.T) {
 	case <-time.After(10 * time.Second):
 		t.Fatal("never received final log message")
 	}
-	assert.Equal(t, []string{"ready\n", "echo hello\n", "echo there\n"}, g.Msgs)
+	assert.Equal(t, []string{"ready\n", "echo hello\n", "echo there\n"}, g.Msgs())
 
 	terminateContainerOnEnd(t, ctx, c)
 }
@@ -443,8 +457,8 @@ func Test_StartLogProductionStillStartsWithTooLowTimeout(t *testing.T) {
 	ctx := context.Background()
 
 	g := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 
@@ -475,8 +489,8 @@ func Test_StartLogProductionStillStartsWithTooHighTimeout(t *testing.T) {
 	ctx := context.Background()
 
 	g := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 
@@ -518,10 +532,11 @@ func Test_MultiContainerLogConsumer_CancelledContext(t *testing.T) {
 
 	// Context with cancellation functionality for simulating user interruption
 	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel() // Ensure it gets called.
 
 	first := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 
@@ -555,8 +570,8 @@ func Test_MultiContainerLogConsumer_CancelledContext(t *testing.T) {
 	require.NoError(t, err)
 
 	second := TestLogConsumer{
-		Msgs:     []string{},
-		Done:     make(chan bool),
+		msgs:     []string{},
+		Done:     make(chan struct{}),
 		Accepted: devNullAcceptorChan(),
 	}
 
@@ -592,7 +607,7 @@ func Test_MultiContainerLogConsumer_CancelledContext(t *testing.T) {
 	// Handling the termination of the containers
 	defer func() {
 		shutdownCtx, shutdownCancel := context.WithTimeout(
-			context.Background(), 60*time.Second,
+			context.Background(), 10*time.Second,
 		)
 		defer shutdownCancel()
 		_ = c.Terminate(shutdownCtx)
@@ -604,8 +619,8 @@ func Test_MultiContainerLogConsumer_CancelledContext(t *testing.T) {
 
 	// We check log size due to context cancellation causing
 	// varying message counts, leading to test failure.
-	assert.GreaterOrEqual(t, len(first.Msgs), 2)
-	assert.GreaterOrEqual(t, len(second.Msgs), 2)
+	assert.GreaterOrEqual(t, len(first.Msgs()), 2)
+	assert.GreaterOrEqual(t, len(second.Msgs()), 2)
 
 	// Restore stderr
 	w.Close()

